add Shadow Keys

hammo92 opened the pull request:
The functionality is working as expected, as far as I can tell, might benefit from more tests and probably some better terminology in the comments.

hammo92 committed:
add Shadow Keys

loreanvictor commented:
wow! this is brilliant, I will take a look asap. kudos!

hammo92 commented:
I think it might be better to replace:
`shadowKeys().map(async shadowKeys => await data.remove(this.primary(shadowKeys)))`
with:
`shadowKeys().map(async keys => await data.remove(this.primary(keys)))`
to make it a little clearer.

hammo92 committed:
Removed unneccessary optional chaining

hammo92 committed:
Added tests to complete coverage

hammo92 committed:
clean up code, remove unnecessary async

Better code formatting

Remove async from map, not necessary as method is already a promise.

hammo92 committed:
Revert "clean up code, remove unnecessary async"

This reverts commit 0e039c41fe01996160bee75aa9de775dc59066ce.

hammo92 committed:
Improve formatting, remove unnecessary async

Better code formatting

Remove async from map, not necessary as method is already a promise

loreanvictor commented:
@hammo92 could you also re-merge main into the branch to resolve the conflict?

hammo92 committed:
Merge branch 'main' into shadow_keys

hammo92 commented:
> @hammo92 could you also re-merge main into the branch to resolve the conflict?

Sorted, but not sure about the Codacy error. I'm guessing that's not on my side?

loreanvictor commented on src/model.ts:
```diff
@@ -67,6 +69,14 @@ export abstract class Model<T extends Model<T>> {
+   * for each set of shadowKeys, each shadowKey list must have a primary, secondary keys must have separate labels.
+   *
+   */
+  shadowKeys?(): Exact[][]
```
shouldn't this method also be marked as `abstract`?

hammo92 replied:
Marking the method as abstract would make it required as far as I can tell. Unless this is the intended implementation?

loreanvictor replied:
Good point.

loreanvictor commented on src/model.ts:
```diff
@@ -90,23 +111,47 @@ export abstract class Model<T extends Model<T>> {
+
+    if (this.shadowKeys) {
+      const shadowKeys = this.shadowKeys()
+      const shadowPrimaries = shadowKeys.map((shadowkeys) =>
```
Here the name `shadowKeys` is used twice, the inner parameter _shadowing_ (no pun intended) the upper scope variable. By default, eslint would warn about this, but since the inner parameter name breaks the convention of being camel case, the warning is bypassed.

That warning however is enforced for code clarity. I feel in this case (and similar cases) we should also use a different name (and honor the camel case convention). For example, in this case the parameter name `keySet` would more clearly reflect the object that is being worked on by the inner function.

loreanvictor commented on src/model.ts:
```diff
@@ -90,23 +111,47 @@ export abstract class Model<T extends Model<T>> {
+        this.primary(shadowkeys)
+      )
+      await Promise.all([
+        ...this.__shadowSnapshots.map((shadowSnapshot, index) => {
```
This code works on the assumption that the `shadowKeys()` method always returns the same number of key sets and the key sets at the same index always correspond to the same primary. This assumption, however, is not true: the `shadowKeys()` might return a changing number of key sets (e.g. having some `tags` field and mapping each tag to a shadow key set).

To resolve this, the check could be changed to whether some primary is missing from the stored primaries in the snapshot, e.g.

```ts
        ...this.__shadowSnapshots.map((shadowSnapshot) => {
          if (!shadowPrimaries.includes(shadowSnapshot)) {
            return data.remove(shadowSnapshot)
          }
        }),
```

hammo92 replied:
neater solution as well

loreanvictor commented on src/model.ts:
```diff
@@ -90,23 +111,47 @@ export abstract class Model<T extends Model<T>> {
     await data.remove(this.primary())
+    if (this.shadowKeys) {
+      await Promise.all(
+        this.shadowKeys().map((shadowKeys) =>
```
While this does not strictly violate the variable shadowing rule, I suspect it would also be more clearer if the inner parameter here was called `keySet` or something in those lines as well.

loreanvictor commented on src/model.ts:
```diff
@@ -76,10 +86,21 @@ export abstract class Model<T extends Model<T>> {
     return pk.query()
   }
 
+  secondaries(keys: Exact[]) {
```
This is a nice utility method, it would also be pretty nice if it could handle an optional `keys` argument similar to `primary()` method.

hammo92 replied:
I'm not sure when this would get called, assuming you pass the same keys to the secondary method as the primary there will always be a primary key in the keys array. Would the purpose be to not run the function if the only key in the array is a primary key?

loreanvictor commented on src/test/model.test.ts:
```diff
@@ -30,20 +41,45 @@ class M extends Model<M> {
+  }
+}
+
+class O extends Model<O> {
```
I suspect a test model here with variable number of shadow key sets would be pretty nice in covering that particular use-case as well.

loreanvictor reviewed (changes requested):
Thanks again for the great work! There is only one functional issue and a few stylish issues with the code which also seem pretty straight-forward to fix! However this implementation currently side-steps the potential limit on number of shadow key sets in a model (see the _Considerations_ section in [the proposal](https://github.com/loreanvictor/serverless-cloud-data-utils/issues/3#considerations)). Such limitation could of course be added in the future, but that would break backwards compatibility, so it would be nice to ponder it on the first iteration as well.

hammo92 commented:
> Thanks again for the great work! There is only one functional issue and a few stylish issues with the code which also seem pretty straight-forward to fix! However this implementation currently side-steps the potential limit on number of shadow key sets in a model (see the _Considerations_ section in [the proposal](https://github.com/loreanvictor/serverless-cloud-data-utils/issues/3#considerations)). Such limitation could of course be added in the future, but that would break backwards compatibility, so it would be nice to ponder it on the first iteration as well.

What do you think would be a good limit for the number of shadowKeys?

loreanvictor commented:
> > Thanks again for the great work! There is only one functional issue and a few stylish issues with the code which also seem pretty straight-forward to fix! However this implementation currently side-steps the potential limit on number of shadow key sets in a model (see the _Considerations_ section in [the proposal](https://github.com/loreanvictor/serverless-cloud-data-utils/issues/3#considerations)). Such limitation could of course be added in the future, but that would break backwards compatibility, so it would be nice to ponder it on the first iteration as well.
> 
> What do you think would be a good limit for the number of shadowKeys?

I am completely blind to any potential parallels / benchmarks, but gut feeling I feel 5 should be a nice limit. It is also highly dependent on potential use-cases, do you have any use-case in mind that would require more than a primary key set and 5 extra key sets? (Keeping in mind that we can always add the _unsafe_ method for unbounded number of key sets).

hammo92 commented:
> > > Thanks again for the great work! There is only one functional issue and a few stylish issues with the code which also seem pretty straight-forward to fix! However this implementation currently side-steps the potential limit on number of shadow key sets in a model (see the _Considerations_ section in [the proposal](https://github.com/loreanvictor/serverless-cloud-data-utils/issues/3#considerations)). Such limitation could of course be added in the future, but that would break backwards compatibility, so it would be nice to ponder it on the first iteration as well.
> > 
> > 
> > What do you think would be a good limit for the number of shadowKeys?
> 
> I am completely blind to any potential parallels / benchmarks, but gut feeling I feel 5 should be a nice limit. It is also highly dependent on potential use-cases, do you have any use-case in mind that would require more than a primary key set and 5 extra key sets? (Keeping in mind that we can always add the _unsafe_ method for unbounded number of key sets).

5 is probably fine, without running some benchmarks it's hard to know but there are likely better solutions for cases where more keys would be useful. An example use case would likely be something like a group chat, where the shadowKeys would make it easier to get the conversation by userID.

hammo92 committed:
requested changes to shadow key pull request

hammo92 committed:
Revert "requested changes to shadow key pull request"

This reverts commit e427237521a5f8a557bb22b91ea79fa3f789e3e6.

hammo92 committed:
make requested changes to pull request

hammo92 committed:
remove unused import

hammo92 commented:
@loreanvictor, No idea how I pushed the completely wrong commit there, but this one should have all the fixes. There may be a cleaner solution for limiting the shadow keys, have a look and let me know.

loreanvictor commented on src/model.ts:
```diff
@@ -90,23 +123,52 @@ export abstract class Model<T extends Model<T>> {
+    if (this.shadowKeys) {
+      if (
+        this.shadowKeys().length > MAX_SHADOW_KEYS &&
+        !this.unsafeShadowKeysUnbounded?.()
```
I think this would not work as intended:

- This will ignore the result from `this.unsafeShadowKeysUnbounded()`, while when trying to get an unbounded number of shadow keys, I guess the intuitive thing would be to just override this function.

- This will assume that there are no shadow keys if `shadowKeys()` method is not defined. However, one might only defined `unsafeShadowKeysUnbounded()` instead. In fact, I feel the intuitive default is to override only one of the two and not both (you either define shadow keys normally, or you choose the unsafe, unbounded way).

My recommendation would be to have another intermediary method who figures out the shadow keys (and checks the bounds as well), and calling that function in `save()`, separating the _bounding_ logic into another method.

```ts
resolveShadowKeys() {
  if (this.unsafeShadowKeysUnbounded) {
    return this.unsafeShadowKeysUnbounded()
  } else if (this.shadowKeys) {
    const res = this.shadowKeys()
    if (res.length > MAX_SHADOW_KEYS) {
      throw ...
    }
    
    return res
  }
}
```

hammo92 replied:
So I think the way I have implemented the unsafeShadowKeysUnbounded is different to how you expected, essentially it is just a flag to say to ignore the MAX_SHADOW_KEYS restriction, the shadow keys are always set in the shadowKeys method. I was thinking that it might make more sense to add an options method instead:
```
setOptions() {
    return {
      shadowKeysLimit: false
    }
 }
```

loreanvictor replied:
In that case I would highly recommend switching to having one overriden method whose name acts as the flag (either `shadowKeys()` or `unsafeShadowKeysUnbounded()`), instead of potentially two overriden methods where one is a flag determining how the other is treated (both `shadowKeys()` and `shadowKeysBoundFlag()` or `setOptions()` or another method). 


In the first pattern (which I believe is [somewhat familiar already](https://reactjs.org/docs/react-component.html#unsafe_componentwillupdate)), the behavior is isolated and clear (here we have shadow keys, or here we have unbounded shadow keys which is potentially unsafe). In the second pattern, you would need to look at another part of the code (or potentially some other class entirely) to understand the behavior, which increases complexity.

hammo92 replied:
Would it be better to change `unsafeShadowKeysUnbounded` to `UNSAFE_shadowKeysUnbounded`? 

If keys are provided for both `unsafeShadowKeysUnbounded` and `shadowKeys` methods, would it make sense to throw an error, or just document that in this case any keys provided to the `shadowKeys` method will be ignored?

loreanvictor replied:
Good idea, `UNSAFE_shadowKeysUnbounded()` better communicates the underlying unsafety, so it indeed is a better choice.

I would recommend throwing an error for now, since that restriction can be relaxed later without breaking backwards compatibility, but not the other way around.

loreanvictor commented on src/model.ts:
```diff
@@ -76,10 +96,23 @@ export abstract class Model<T extends Model<T>> {
+   * the previous entry in the database will be deleted first, this also runs for primary shadow keys.
    *
    */
   async save() {
```
If I am not mistaken, this function is also a bit buggy (it is not a newly introduced bug, this bug is present in the current published version as well): The snapshot is NOT updated, so if you change the model in a way that results in change of primary:

- upon first save, the previous instance is deleted.
- upon second save, the previous instance is again deleted (while it probably doesn't exist).

The same issue is also present for shadow primaries. Perhaps all primaries should be updated upon saving?

hammo92 replied:
Does this not make the snapshot redundant, is there a reason to not update the snapshot when the primary key updates?

loreanvictor replied:
> is there a reason to not update the snapshot when the primary key updates?

Yes. The snapshot needs to be updated after the model is _saved to the database_, which happens way after the primary key updates. When saving to db, you definitely need the snapshot in order to clean up the previous object (if need be), which means the information must be preserved through updates to the primary key before the instance is saved.

```ts
const bob = await indexBy(Name).exact('Bob').get(Person)

bob.name = 'Rob' // 👉 primary key is updated
bob.name = 'Jacob' // 👉 primary key is updated

bob.save()
// ☝️Here we need to remember the original name, 'Bob',
//   as we need to delete the record with name 'Bob' from db.
//   This means we had to preserve the "snapshot" (the original name)
//   despite having updated the primary key.
```

hammo92 replied:
Sure, I understand the purpose of the snapshot. Is this happening if the save function is called twice on the same instance? 
```
const bob = await indexBy(Name).exact('Bob').get(Person)

bob.name = 'Rob' // 👉 primary key is updated
bob.name = 'Jacob' // 👉 primary key is updated

bob.save()

bob.name = "John"

bob.save()
```

hammo92 replied:
Sure, I was thinking that the snapshot could be updated in the comparison after deleting the existing entry by primary, so that the subsequent save call would have the correct snapshot
```
if (this.__snapshot && primary !== this.__snapshot) {
      await data.remove(this.__snapshot)
      this.__snapshot = primary
}
```

loreanvictor replied:
Exactly. And I suspect the same should go for primary snapshots (they need to get updated after all deletions and updates are conducted to the db).

hammo92 replied:
okay, cool. That's an easy enough fix

loreanvictor commented on src/model.ts:
```diff
@@ -90,23 +123,52 @@ export abstract class Model<T extends Model<T>> {
+   * Will delete the model and any shadows from the database.
    *
    */
   async delete() {
```
This method also has a similar bug to `save()`. When deleting a model, the snapshots (both the main and shadow snapshots) are not updated (should be set to `undefined` I suppose?). To re-iterate, this is not a newly introduced bug, but one that exists in current version as well, but it also affects shadow primaries.

loreanvictor commented:
> @loreanvictor, No idea how I pushed the completely wrong commit there, but this one should have all the fixes. There may be a cleaner solution for limiting the shadow keys, have a look and let me know.

@hammo92 the solution is fine on its own from a cleanliness perspective, however it has marginal functional issues (I have commented on the code). I think after these changes we should merge and go to prod, as the changes seem robust enough.

hammo92 committed:
Requested changes to shadowKeys request

snapshots update on save and delete
UNSAFE_shadowKeysUnbounded method altered to return keys
resolveShadowKeys method added
updateShadowSnapshots method added
test added for using both shadowKeys methods together

loreanvictor commented:
@hammo92 merged, thanks for the great work!
